include/nektar_interface/projection/tri.hpp

```diff
@@ -0,0 +1,120 @@
+#pragma once
+#include "algorithm_types.hpp"
+#include "basis/basis.hpp"
+#include "constants.hpp"
+// #include "device_data.hpp"
+#include "restrict.hpp"
+#include "unroll.hpp"
+#include "util.hpp"
+namespace NESO::Project {
+
+namespace Private {
+struct eTriangleBase {
+private:
+public:
+  static constexpr Nektar::LibUtilities::ShapeType shape_type =
+      Nektar::LibUtilities::eTriangle;
+  static constexpr int dim = 2;
+  template <typename T>
+  inline static void NESO_ALWAYS_INLINE
+  loc_coord_to_loc_collapsed(T const xi0, T const xi1, T &eta0, T &eta1) {
+    eta0 = Util::Private::collapse_coords(xi0, xi1);
+    eta1 = xi1;
+  }
+  template <int nmode> static auto NESO_ALWAYS_INLINE get_ndof() {
+    return nmode * (nmode + 1) / 2;
+  }
+};
+} // namespace Private
+
+template <typename Algorithm>
+struct eTriangle : public Private::eTriangleBase {};
+
+template <> struct eTriangle<ThreadPerDof> : public Private::eTriangleBase {
+  using algorithm = ThreadPerDof;
+
+private:
+  // solving for
+  //(nmode + 1 + (nmode +1 - dof))*(dof + 1)/2 = X;
+  // if nmode==4
+  // then
+  // 0,1,2,3,4 -> 0
+  // 5,6,7,8   -> 1
+  // 9,10,11   -> 2
+  // 12,13     -> 3
+  // 14        -> 4
+  // i.e. mapping from dof -> index in emodA array
+  template <int nmode>
```

Comment:
Can we avoid this floating point maths by computing a lookup table (static maybe to keep the memory footprint down)?